README.md
# DLA Simulator
## Introduzione
DLA Simulator è un programma in C per simulare l'aggregazione di particelle che si muovono seguendo un moto Browniano in una superficie 2D. \
Presentiamo più implementazioni, una single core, una multi thread usando la libreria pthread e una multi thread usando OpenMP. \
Inoltre le nostre implementazioni prevedono il render delle fasi della simulazioni per mostrare come si aggregano le particelle.
## Come avviarlo
**Il progetto è valido solo se fatto girare su Linux, non si assicura il suo funzionamento su altre piattaforme visto l'utilizzo di librerie multi thread ed altro** \
Dopo aver clonato la repository è possibile avviare il programma a patto che le dipendenze vengano rispettate. (Vedi sezione <a href="Dipendenze"> Dipendenze</a>)
Quindi per compilarlo basta eseguire il seguente comando:

    gcc -o dla_single_thread.out dla_single_thread.c -lgd

Dove _-lgd_ è l'opzione necessaria per specificare la libreria GD che usiamo per il render delle immagini. \
\
Per le versioni multi thread va aggiunta l'opzione:

 - `-fopenmp` _per l'implementazione con OpenMP_
 - `-pthread` _per l'implementazione con pthread_

Per avviare la versione single thread basta eseguire il seguente comando e settare i giusti argomenti:

 - n,m _grandezza matrice, righe e colonne rispettivamente_
 - num_particles _numero di particelle da istanziare_
 - x,y _posizione del seed con le rispettive coordinate nella matrice_ [***opzionale***] \ Se non specificato sarà generato casualmente all'interno della superficie.

`./dla_single_thread.out n,m num_particles x,y`

Per le versioni multi thread:

 - OpenMP - `./dla_openmp.out n,m num_particles [-OPTIONS]]` 
 - pthread - `./dla_pthread.out n,m num_particles [-OPTIONS]` \

Dove n,m sono le dimensioni della matrice, num_particles è il numero di particelle da istanziare. \
Le opzioni sono:
 - -n _numero di thread da usare_ [default 4]
 - -s _posizione del seed_ [default generato casualmente]
 - -t _orizzonte di simulazione_ [default 1000]

## Come funziona
Il programma è molto semplice, le tre implementazioni differiscono di poco, l'idea di fondo è sempre la stessa.
L' inizializzazione prevede il salvataggio dei dati passati come argomento al programma, l'allocazione della memoria per la matrice, che viene allocata e riempita interamente da 0, l'allocazione della memoria per la lista di particelle, queste vengono generate con posizione casuale. Dopo di che inizia la simulazione vera e propria. \
Tutta la simulazione è scandita da intervalli (l'unità di tempo la lasciamo scegliere all'utente), in ognuno di questi intervalli si scorre tutta la lista di particelle, e per ogni particella si eseguono due operazioni:

 1. Si controllano le celle adiacenti alla particella, nel caso sia presente una particella aggregata (o il seed iniziale) allora la particella in questione si aggregherà a sua volta e non verrà più considerata, in caso contrario si va alla seconda operazione.
 2. Si muove la particella di un fattore casuale in una direzione casuale cercando di simulare al meglio un moto Browniano. Quindi si ricomincia dal punto 1. 

Al termine dell'orizzonte di simulazione scelto viene salvato il risultato sottoforma di immagine in formato jpg, dopo di che viene liberata tutta la memoria allocata e nel caso delle implementazioni multi thread vengono eseguite le dovute operazioni di finalizzazione.

## Dipendenze
Per il render delle immagini in C abbiamo usato: [GD Graphics Library (libgd.github.io)](https://libgd.github.io/)
Per installarla su Ubuntu:
`sudo apt install libgd-dev`


times/history.txt
[(10, 10, 30, 4, 100), (10, 10, 30, 4, 1000), (100, 100, 400, 4, 100), (100, 100, 400, 4, 1000), (100, 100, 400, 4, 10000), (100, 100, 400, 8, 10000), (100, 100, 700, 4, 1000), (1000, 1000, 300000, 4, 1000), (1000, 1000, 300000, 8, 1000), (1000, 1000, 600000, 4, 1000), (1000, 1000, 600000, 8, 1000)] 
 [0.0, 0.0, 0.0, 0.0, 0.05, 0.12708333333333333, 0.18214285714285713, 4.7453125, 11.33611111111111, 23.64875, 38.798863636363635] 
 
[(100, 100, 400, 0, 1000, 'dla_single_thread'), (100, 100, 400, 4, 1000, 'dla_pthread'), (100, 100, 400, 4, 1000, 'dla_openmp'), (100, 100, 400, 8, 1000, 'dla_pthread'), (100, 100, 400, 8, 1000, 'dla_openmp'), (100, 100, 400, 16, 1000, 'dla_pthread'), (100, 100, 400, 16, 1000, 'dla_openmp'), (100, 100, 700, 0, 1000, 'dla_single_thread'), (100, 100, 700, 4, 1000, 'dla_pthread'), (100, 100, 700, 4, 1000, 'dla_openmp'), (100, 100, 700, 8, 1000, 'dla_pthread'), (100, 100, 700, 8, 1000, 'dla_openmp'), (100, 100, 700, 16, 1000, 'dla_pthread'), (100, 100, 700, 16, 1000, 'dla_openmp')] 
 [0.04, 0.04, 0.04, 0.02, 0.08, 0.05, 0.11, 0.22, 0.11, 0.15, 0.12, 0.16, 0.13, 0.17] 
 
[(1000, 1000, 300000, 0, 1000, 'dla_single_thread'), (1000, 1000, 300000, 4, 1000, 'dla_pthread'), (1000, 1000, 300000, 4, 1000, 'dla_openmp'), (1000, 1000, 300000, 8, 1000, 'dla_pthread'), (1000, 1000, 300000, 8, 1000, 'dla_openmp'), (1000, 1000, 300000, 16, 1000, 'dla_pthread'), (1000, 1000, 300000, 16, 1000, 'dla_openmp'), (1000, 1000, 600000, 0, 1000, 'dla_single_thread'), (1000, 1000, 600000, 4, 1000, 'dla_pthread'), (1000, 1000, 600000, 4, 1000, 'dla_openmp'), (1000, 1000, 600000, 8, 1000, 'dla_pthread'), (1000, 1000, 600000, 8, 1000, 'dla_openmp'), (1000, 1000, 600000, 16, 1000, 'dla_pthread'), (1000, 1000, 600000, 16, 1000, 'dla_openmp')] 
 [73.95, 118.72, 148.87, 113.53, 141.57, 125.68, 147.69, 328.45, 209.75, 238.57, 238.93, 270.59, 270.77, 300.42] 
 
[(1000, 1000, 300000, 0, 1000, 'dla_single_thread'), (1000, 1000, 300000, 4, 1000, 'dla_pthread'), (1000, 1000, 300000, 4, 1000, 'dla_openmp'), (1000, 1000, 300000, 8, 1000, 'dla_pthread'), (1000, 1000, 300000, 8, 1000, 'dla_openmp'), (1000, 1000, 300000, 16, 1000, 'dla_pthread'), (1000, 1000, 300000, 16, 1000, 'dla_openmp')] 
 [32.13, 44.46, 56.71, 37.54, 46.55, 38.16, 45.13] 
 
[(1000, 1000, 700000, 0, 1000, 'dla_single_thread'), (1000, 1000, 700000, 4, 1000, 'dla_pthread'), (1000, 1000, 700000, 4, 1000, 'dla_openmp'), (1000, 1000, 700000, 8, 1000, 'dla_pthread'), (1000, 1000, 700000, 8, 1000, 'dla_openmp'), (1000, 1000, 700000, 16, 1000, 'dla_pthread'), (1000, 1000, 700000, 16, 1000, 'dla_openmp')] 
 [62.55, 88.27, 112.7, 75.74, 93.87, 77.68, 91.72] 
 
[(1000, 1000, 300000, 0, 1000, 'dla_single_thread'), (1000, 1000, 300000, 4, 1000, 'dla_pthread'), (1000, 1000, 300000, 4, 1000, 'dla_openmp'), (1000, 1000, 300000, 8, 1000, 'dla_pthread'), (1000, 1000, 300000, 8, 1000, 'dla_openmp'), (1000, 1000, 300000, 16, 1000, 'dla_pthread'), (1000, 1000, 300000, 16, 1000, 'dla_openmp')] 
 [30.82, 43.55, 55.98, 40.71, 53.13, 48.13, 60.63] 
 
[(1000, 1000, 700000, 0, 1000, 'dla_single_thread'), (1000, 1000, 700000, 4, 1000, 'dla_pthread'), (1000, 1000, 700000, 4, 1000, 'dla_openmp'), (1000, 1000, 700000, 8, 1000, 'dla_pthread'), (1000, 1000, 700000, 8, 1000, 'dla_openmp'), (1000, 1000, 700000, 16, 1000, 'dla_pthread'), (1000, 1000, 700000, 16, 1000, 'dla_openmp')] 
 [62.8, 29.69, 25.58, 28.62, 25.73, 28.21, 25.98] 
 
[(1000, 1000, 700000, 0, 1000, 'dla_single_thread'), (1000, 1000, 700000, 4, 1000, 'dla_pthread'), (1000, 1000, 700000, 4, 1000, 'dla_openmp'), (1000, 1000, 700000, 8, 1000, 'dla_pthread'), (1000, 1000, 700000, 8, 1000, 'dla_openmp'), (1000, 1000, 700000, 16, 1000, 'dla_pthread'), (1000, 1000, 700000, 16, 1000, 'dla_openmp')] 
 [64.09, 26.91, 24.65, 29.3, 24.92, 29.52, 25.28] 
 
[(1000, 1000, 400000, 0, 10000, 'dla_single_thread'), (1000, 1000, 400000, 4, 10000, 'dla_pthread'), (1000, 1000, 400000, 4, 10000, 'dla_openmp'), (1000, 1000, 400000, 8, 10000, 'dla_pthread'), (1000, 1000, 400000, 8, 10000, 'dla_openmp'), (1000, 1000, 400000, 16, 10000, 'dla_pthread'), (1000, 1000, 400000, 16, 10000, 'dla_openmp')] 
 [128.68, 55.06, 48.64, 54.05, 49.18, 54.29, 50.52] 
 
[(1000, 1000, 40, 0, 100, 'dla_single_thread')] 
 [0.0] 
 
[(1000, 1000, 40, 0, 100, 'dla_single_thread')] 
 [0.0] 
 
[(1000, 1000, 700000, 0, 10000, 'dla_single_thread'), (1000, 1000, 700000, 4, 10000, 'dla_pthread'), (1000, 1000, 700000, 4, 10000, 'dla_openmp'), (1000, 1000, 700000, 8, 10000, 'dla_pthread'), (1000, 1000, 700000, 8, 10000, 'dla_openmp'), (1000, 1000, 700000, 16, 10000, 'dla_pthread'), (1000, 1000, 700000, 16, 10000, 'dla_openmp')] 
 [108.29, 45.52, 41.52, 46.56, 41.83, 46.31, 43.14] 
 
[(1000, 1000, 700000, 0, 10000, 'dla_single_thread'), (1000, 1000, 700000, 4, 10000, 'dla_pthread'), (1000, 1000, 700000, 4, 10000, 'dla_openmp'), (1000, 1000, 700000, 8, 10000, 'dla_pthread'), (1000, 1000, 700000, 8, 10000, 'dla_openmp'), (1000, 1000, 700000, 16, 10000, 'dla_pthread'), (1000, 1000, 700000, 16, 10000, 'dla_openmp')] 
 [114.62, 51.41, 44.27, 49.17, 44.5, 49.0, 45.22] 
 
[(1000, 1000, 400000, 0, 10000, 'dla_single_thread'), (1000, 1000, 400000, 4, 10000, 'dla_pthread'), (1000, 1000, 400000, 4, 10000, 'dla_openmp'), (1000, 1000, 400000, 8, 10000, 'dla_pthread'), (1000, 1000, 400000, 8, 10000, 'dla_openmp'), (1000, 1000, 400000, 16, 10000, 'dla_pthread'), (1000, 1000, 400000, 16, 10000, 'dla_openmp')] 
 [157.89, 62.68, 58.53, 64.75, 59.3, 65.62, 59.86] 
 


pytest/myplot.py
#!/bin/python3
''''
Speedup del sistema parallelo: Sp = T1/Tp
Efficienza del sistema parallelo: Ep = Sp/p
Overhead del sistema parallelo: Oh = pTp - T1

p = num_threads

- Lo speedup misura la riduzione del tempo di esecuzione rispetto
all'algoritmo su un processore. Nel caso ideale Sp = p;

- L'efficienza misura quanto l'algoritmo sfrutta il parallelismo del
calcolatore. Nel caso ideale Efficienza = 1;

- L'overhead misura quanto lo speedup differisce da quello ideale, ovvero
rappresenta un'indicazione sullo spreco delle risorse di calcolo. 
Idealmente Oh = 0.
'''
import numpy as np
import matplotlib.pyplot as plt

EXE_FILENAMES = ("dla_single_thread", "dla_pthread", "dla_openmp")


'''
args = (n,m,num_particles, threads, n1, m1, num_particles1, threads1, ...)
'''
def calc_value():
        return



def getPlot(elapsedTimes: list, args: tuple[str]):
    #SPEEDUP
    speedup = np.array([])
    for time in elapsedTimes:
        speedup = np.append(speedup, np.array(
            [round(time[0]/data, 4) for data in time[1:]]))
        
    percent = [int(config[2]/(config[0]*config[1])*100) for config in args]
    plt.figure(figsize=(10, 8))
    x_pos = np.array([config[2] for config in args[1:-1]])
    y_pos = tuple([1/x for x in speedup])
    #plt.plot(x_pos, y_pos, data=speedup, marker='o',
     #        color='b')
    print(percent)
    
    _, ax = plt.subplots()
    ax.plot(x_pos, y_pos, data=speedup, marker='o', color='b')
    
    for i, txt in enumerate(x_pos):
        ax.annotate(f"{percent[i]}%", (x_pos[i], y_pos[i]))
    plt.xticks(x_pos, fontsize=8, ha='right')
    plt.yticks(y_pos, fontsize=8)
    plt.ylabel('Speedup')
    plt.xlabel('Numero di particelle')
    plt.title('DLA \n Speedup in funzione al numero di particelle')
    plt.savefig('test_plot.png')
    return


def getBar(configurations : list, single_times: list, pthread_times: list, openmp_times: list):

	x = np.array([config[2] for config in configurations])
	s_time = np.array([time[0] for time in single_times])
	p_time = np.array([time[0] for time in pthread_times])
	o_time = np.array([time[0] for time in openmp_times])

	offset = 0.2
	for i in range(0, len(2)):
			#single
			plt.bar(x-offset, s_time[i])
			#pthread
			plt.bar(x, p_time[i])
			#openmp
			plt.bar(x+offset, o_time[i])
			
	plt.savefig('test_bar.png')
	return

def inList(s : str):
    s = s.replace('\t', ' ')
    s = s.replace(',', '.')
    s = ' '.join(s.split())
    speedup = [float(y) for y in s.split(' ')]
    p4, p8, p16, o4, o8, o16 = [],[],[],[],[],[]
    for i in range(0, len(speedup), 6):
        p4.append(speedup[i])
        o4.append(speedup[i+1])
        p8.append(speedup[i+2])
        o8.append(speedup[i+3])
        p16.append(speedup[i+4])
        o16.append(speedup[i+5])
        
    speedup = [p4, o4, p8, o8, p16, o16]
    return speedup


def getPlot2(speedup : list , x_pos : list, percent : list):
    speedup_np = np.array(speedup)
    print(speedup_np)
    m = int(np.max(speedup_np))
    y_pos = np.arange(m-0.1, m+1, 0.05)
    x_pos_np = np.array(x_pos)
	
	
    _, ax = plt.subplots()
 
    #p4
    plt.plot(x_pos, speedup_np[0], marker='o', color='b') 
    #o4
    plt.plot(x_pos, speedup_np[1], marker='o', color='r')
    #p8
    plt.plot(x_pos, speedup_np[2], marker='o', color='g')
    #o8
    plt.plot(x_pos, speedup_np[3], marker='o', color='y')
    #p16
    plt.plot(x_pos, speedup_np[4], marker='o', color='c')
    #o16
    plt.plot(x_pos, speedup_np[5], marker='o', color='m')
        
    
    
    plt.xticks(x_pos, fontsize=8, ha='center')
    plt.yticks(y_pos, fontsize=8)
    
    plt.legend(['pthread 4 threads', 'openMP 4 threads',
                'pthread 8 threads', 'openMP 8 threads',
                'pthread 16 threads', 'openMP 16 threads'])

    plt.ylabel('Speedup')
    plt.xlabel('Numero di particelle')
    plt.title('DLA \n Speedup in funzione al numero di particelle')
    plt.savefig('test_plot.png')
    return


if __name__ == '__main__':
    configurations = [
        # (n, m, num_particles, num_threads, iters, filename)
        (100, 100, 400, 0, 1000, EXE_FILENAMES[0]),
        (100, 100, 400, 4, 1000,  EXE_FILENAMES[1]),
        (100, 100, 400, 4, 1000, EXE_FILENAMES[2]),
        (100, 100, 400, 8, 1000, EXE_FILENAMES[1]),
        (100, 100, 400, 8, 1000, EXE_FILENAMES[2]),
        (100, 100, 400, 16, 1000, EXE_FILENAMES[1]),
        (100, 100, 400, 16, 1000, EXE_FILENAMES[2]),
        (100, 100, 700, 0, 1000, EXE_FILENAMES[0]),
        (100, 100, 700, 4, 1000, EXE_FILENAMES[1]),
        (100, 100, 700, 4, 1000, EXE_FILENAMES[2]),
        (100, 100, 700, 8, 1000, EXE_FILENAMES[1]),
        (100, 100, 700, 8, 1000, EXE_FILENAMES[2]),
        (100, 100, 700, 16, 1000, EXE_FILENAMES[1]),
        (100, 100, 700, 16, 1000, EXE_FILENAMES[2]),
    ]
	
    s = '2,22	2,12	2,35	2,33	2,35	2,28 2,42	2,48	2,43	2,48	2,42	2,44 2,40	2,55	2,38	2,48	2,38	2,51 2,21	2,49	2,11	2,45	2,18	2,38 2,15	2,53	1,93	2,53	2,02	2,48'
    speedup = inList(s)
    num_p = [i*100000 for i in range(1,10,2)]
    getPlot2(speedup, num_p, [])
	
    pass
